fractal_tasks_core/tasks/import_ome_zarr.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2022 (C) Friedrich Miescher Institute for Biomedical Research and
+# University of Zurich
+#
+# Original authors:
+# Tommaso Comparin <tommaso.comparin@exact-lab.it>
+#
+# This file is part of Fractal and was originally developed by eXact lab S.r.l.
+# <exact-lab.it> under contract with Liberali Lab from the Friedrich Miescher
+# Institute for Biomedical Research and Pelkmans Lab from the University of
+# Zurich.
+"""
+Task to import an existing OME-Zarr.
+"""
+import logging
+from pathlib import Path
+from typing import Any
+from typing import Sequence
+
+import dask.array as da
+import zarr
+from pydantic.decorator import validate_arguments
+
+from fractal_tasks_core.lib_ngff import detect_ome_ngff_type
+from fractal_tasks_core.lib_ngff import NgffImageMeta
+from fractal_tasks_core.lib_regions_of_interest import get_image_grid_ROIs
+from fractal_tasks_core.lib_regions_of_interest import get_single_image_ROI
+from fractal_tasks_core.lib_write import write_table
+
+logger = logging.getLogger(__name__)
+
+
+def _process_single_image(
+    image_path: str,
+    add_image_ROI_table: bool,
+    add_grid_ROI_table: bool,
+    grid_YX_shape: tuple[int, int],
+) -> None:
+    """
+    Validate OME-NGFF metadata and optionally generate ROI tables.
+
+    This task:
+
+    1. Validates OME-NGFF image metadata, via `NgffImageMeta`;
+    2. Optionally generates and writes two ROI tables.
+
+    Args:
+        image_path: Absolute path to the image Zarr group.
+        add_image_ROI_table: Whether to add a `image_ROI_table` table
+            (argument propagated from `import_ome_zarr`).
+        add_grid_ROI_table: Whether to add a `grid_ROI_table` table (argument
+            propagated from `import_ome_zarr`).
+        grid_YX_shape: YX shape of the ROI grid.
+    """
+
+    # Note from zarr docs: `r+` means read/write (must exist)
+    image_group = zarr.open_group(image_path, mode="r+")
+    image_meta = NgffImageMeta(**image_group.attrs.asdict())
+
+    if not (add_image_ROI_table or add_grid_ROI_table):
+        return
+
+    pixels_ZYX = image_meta.get_pixel_sizes_zyx(level=0)
+
+    # Read zarr array
+    dataset_subpath = image_meta.datasets[0].path
+    array = da.from_zarr(f"{image_path}/{dataset_subpath}")
+
+    # Prepare image_ROI_table and write it into the zarr group
+    if add_image_ROI_table:
+        image_ROI_table = get_single_image_ROI(array.shape, pixels_ZYX)
+        write_table(
+            image_group,
+            "image_ROI_table",
+            image_ROI_table,
+            overwrite=False,
+            logger=logger,
+        )
+
+    # Prepare grid_ROI_table and write it into the zarr group
+    if add_grid_ROI_table:
+        grid_ROI_table = get_image_grid_ROIs(
+            array.shape,
+            pixels_ZYX,
+            grid_YX_shape,
+        )
+        write_table(
+            image_group,
+            "grid_ROI_table",
+            grid_ROI_table,
+            overwrite=False,
```

Comment:
Should we allow overwrite of ROI tables?
Scenario: A user imported an OME-Zarr, but chose the wrong `grid_YX_shape`. Now they want to rerun with the correct `grid_YX_shape`, but it would fail due to the overwrite.
=> We can have it as a parameter in the task and use it here

Reply:
> Should we allow overwrite of ROI tables?

Sure, this was an option I also thought about.

Now implemented with https://github.com/fractal-analytics-platform/fractal-tasks-core/pull/557/commits/078c0bf17da5505f879d2b53952d5b7ff6914d64 (defaulting to `False`).